File: rust/open/image_layer.rs
use alloc::{fmt, str::FromStr, string::String, vec::Vec};

use pbf::{BitCast, ProtoRead, ProtoWrite, Protobuf};

// TODO: This could be faster if we don't read in the grid data on parsing but only if the user needs it

/// Track the image type
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
pub enum ImageType {
    /// PNG
    #[default]
    PNG = 0,
    /// JPEG
    JPG = 1,
    /// WebP
    WEBP = 2,
    /// GIF
    GIF = 3,
    /// AVIF
    AVIF = 4,
    /// SVG
    SVG = 5,
    /// BMP
    BMP = 6,
    /// RAW
    RAW = 7,
    /// Unknown
    UNKNOWN = 8,
}
impl BitCast for ImageType {
    fn to_u64(&self) -> u64 {
        (*self) as u64
    }
    fn from_u64(value: u64) -> Self {
        match value {
            0 => ImageType::PNG,
            1 => ImageType::JPG,
            2 => ImageType::WEBP,
            3 => ImageType::GIF,
            4 => ImageType::AVIF,
            5 => ImageType::SVG,
            6 => ImageType::BMP,
            7 => ImageType::RAW,
            8 => ImageType::UNKNOWN,
            _ => panic!("unknown value: {}", value),
        }
    }
}
impl FromStr for ImageType {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_uppercase().as_str() {
            "PNG" => Ok(ImageType::PNG),
            "JPG" | "JPEG" => Ok(ImageType::JPG),
            "WEBP" => Ok(ImageType::WEBP),
            "GIF" => Ok(ImageType::GIF),
            "AVIF" => Ok(ImageType::AVIF),
            "SVG" => Ok(ImageType::SVG),
            "BMP" => Ok(ImageType::BMP),
            "RAW" => Ok(ImageType::RAW),
            "UNKNOWN" => Ok(ImageType::UNKNOWN),
            #[tarpaulin::skip]
            _ => Err("Unknown image type"),
        }
    }
}
impl From<&str> for ImageType {
    fn from(s: &str) -> Self {
        ImageType::from_str(s.to_uppercase().as_str()).unwrap_or(ImageType::PNG)
    }
}
impl fmt::Display for ImageType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let name = match self {
            ImageType::PNG => "PNG",
            ImageType::JPG => "JPG",
            ImageType::WEBP => "WEBP",
            ImageType::GIF => "GIF",
            ImageType::AVIF => "AVIF",
            ImageType::SVG => "SVG",
            ImageType::BMP => "BMP",
            ImageType::RAW => "RAW",
            ImageType::UNKNOWN => "UNKNOWN",
        };
        write!(f, "{}", name)
    }
}

/// Elevation object to read from
#[derive(Debug, Default, Clone, PartialEq)]
pub struct ImageData {
    /// The name of the image
    pub name: String,
    /// The image type
    pub image_type: ImageType,
    /// The image width
    pub width: u32,
    /// The image height
    pub height: u32,
    /// The image data
    pub image: Vec<u8>,
}
impl ImageData {
    /// Create a new ImageData
    pub fn new(
        name: String,
        image_type: ImageType,
        width: u32,
        height: u32,
        image: Vec<u8>,
    ) -> Self {
        Self { name, image_type, width, height, image }
    }
}
impl ProtoRead for ImageData {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            0 => self.image_type = pb.read_varint::<ImageType>(),
            1 => self.width = pb.read_varint(),
            2 => self.height = pb.read_varint(),
            3 => self.image = pb.read_bytes(),
            4 => self.name = pb.read_string(),
            #[tarpaulin::skip]
            _ => panic!("unknown tag: {}", tag),
        }
    }
}
impl ProtoWrite for ImageData {
    fn write(&self, pb: &mut Protobuf) {
        pb.write_varint_field(0, self.image_type);
        pb.write_varint_field(1, self.width);
        pb.write_varint_field(2, self.height);
        pb.write_bytes_field(3, &self.image);
        pb.write_string_field(4, &self.name);
    }
}

File: src/open/imageLayer.ts
import { PbfReader, Pbf as Protobuf } from 'pbf-ts';

/** Image type */
export const enum ImageType {
  /** PNG Image */
  PNG = 0,
  /** JPEG Image */
  JPG = 1,
  /** WEBP Image */
  WEBP = 2,
  /** GIF Image */
  GIF = 3,
  /** AVIF Image */
  AVIF = 4,
  /** SVG Image */
  SVG = 5,
  /** BMP Image */
  BMP = 6,
  /** RAW Image */
  RAW = 7,
  /** Unknown image type */
  UNKNOWN = 8,
}
/** String version of ImageType */
export type ImageTypeString =
  | 'raw'
  | 'png'
  | 'jpg'
  | 'webp'
  | 'gif'
  | 'avif'
  | 'svg'
  | 'bmp'
  | 'unknown';

/** Elevation object to read from */
export class ImageData {
  #pbf: PbfReader;
  name: string = 'default';
  private imagePos: number = 0;
  type: ImageTypeString = 'png';
  width: number = 512;
  height: number = 512;
  /**
   * @param pbf - the pbf protocol we are reading from
   * @param end - the position to stop at
   */
  constructor(pbf: PbfReader, end: number) {
    this.#pbf = pbf;
    pbf.readFields(this.#readImage, this, end);
  }

  /**
   * Reads the image data
   * @returns - the image data
   */
  image(): Uint8Array {
    this.#pbf.pos = this.imagePos;
    return this.#pbf.readBytes();
  }

  /**
   * @param tag - the tag to read
   * @param imageData - the image data to mutate
   * @param pbf - the Protobuf to pull the appropriate data from
   */
  #readImage(tag: number, imageData: ImageData, pbf: PbfReader): void {
    if (tag === 1) imageData.type = fromImageType(pbf.readVarint() as ImageType);
    else if (tag === 2) imageData.width = pbf.readVarint();
    else if (tag === 3) imageData.height = pbf.readVarint();
    else if (tag === 4) imageData.imagePos = pbf.pos;
    else if (tag === 5) imageData.name = pbf.readString();
  }
}

/** Necessary data to encode an image */
export interface ImageDataInput {
  /** The name of the image */
  name: string;
  /** The image type */
  type: ImageTypeString;
  /** The width of the image */
  width: number;
  /** The height of the image */
  height: number;
  /** The raw image data */
  image: Buffer | Uint8Array | ArrayBuffer;
}

/**
 * @param input - the image data & metadata to encode
 * @returns - the encoded tile data
 */
export function writeImageData(input: ImageDataInput): Uint8Array {
  const pbf = new Protobuf();
  const { type, width, height, image } = input;
  pbf.writeVarintField(1, toImageType(type));
  pbf.writeVarintField(2, width);
  pbf.writeVarintField(3, height);
  pbf.writeBytesField(4, image);
  pbf.writeStringField(5, input.name);

  return pbf.commit();
}

/**
 * @param imageType - the enum to convert
 * @returns - the string
 */
export function fromImageType(imageType: ImageType): ImageTypeString {
  if (imageType === ImageType.PNG) return 'png';
  else if (imageType === ImageType.JPG) return 'jpg';
  else if (imageType === ImageType.WEBP) return 'webp';
  else if (imageType === ImageType.GIF) return 'gif';
  else if (imageType === ImageType.AVIF) return 'avif';
  else if (imageType === ImageType.SVG) return 'svg';
  else if (imageType === ImageType.BMP) return 'bmp';
  else if (imageType === ImageType.RAW) return 'raw';
  else if (imageType === ImageType.UNKNOWN) return 'unknown';
  throw new Error('Invalid image type');
}

/**
 * @param imageType - the string to convert
 * @returns - the enum
 */
export function toImageType(imageType: ImageTypeString): ImageType {
  if (imageType === 'png') return ImageType.PNG;
  else if (imageType === 'jpg') return ImageType.JPG;
  else if (imageType === 'webp') return ImageType.WEBP;
  else if (imageType === 'gif') return ImageType.GIF;
  else if (imageType === 'avif') return ImageType.AVIF;
  else if (imageType === 'svg') return ImageType.SVG;
  else if (imageType === 'bmp') return ImageType.BMP;
  else if (imageType === 'raw') return ImageType.RAW;
  else if (imageType === 'unknown') return ImageType.UNKNOWN;
  throw new Error('Invalid image type');
}
